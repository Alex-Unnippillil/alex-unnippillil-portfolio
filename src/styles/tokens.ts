export type TokenName = 'surface' | 'overlay' | 'border' | 'muted' | 'brand';

export type Theme = 'light' | 'dark';

export interface TokenSet {
  surface: string;
  overlay: string;
  border: string;
  muted: string;
  brand: string;
}

export const tokens: Record<Theme, TokenSet> = {
  light: {
    surface: '#ffffff',
    overlay: '#f8f9fa',
    border: '#e2e8f0',
    muted: '#6b7280',
    brand: '#047857',
  },
  dark: {
    surface: '#232931',
    overlay: '#2f343d',
    border: '#4b5563',
    muted: '#9ca3af',
    brand: '#4ecca3',
  },
};

function luminance(hex: string): number {
  const value = hex.replace('#', '');
  const r = parseInt(value.substring(0, 2), 16) / 255;
  const g = parseInt(value.substring(2, 4), 16) / 255;
  const b = parseInt(value.substring(4, 6), 16) / 255;

  const [rr, gg, bb] = [r, g, b].map((v) => (
    v <= 0.03928 ? v / 12.92 : ((v + 0.055) / 1.055) ** 2.4
  ));

  return 0.2126 * rr + 0.7152 * gg + 0.0722 * bb;
}

function contrast(a: string, b: string): number {
  let la = luminance(a);
  let lb = luminance(b);
  if (la < lb) {
    [la, lb] = [lb, la];
  }
  return (la + 0.05) / (lb + 0.05);
}

function validateContrast(set: TokenSet): void {
  const pairs: [TokenName, TokenName][] = [
    ['surface', 'brand'],
    ['surface', 'muted'],
    ['overlay', 'brand'],
    ['overlay', 'muted'],
  ];

  pairs.forEach(([bg, fg]) => {
    const ratio = contrast(set[bg], set[fg]);
    if (ratio < 4.5) {
      // eslint-disable-next-line no-console
      console.warn(`Low contrast between ${bg} and ${fg}: ${ratio.toFixed(2)}`);
    }
  });
}

export function applyTheme(theme: Theme): void {
  const set = tokens[theme];
  const root = document.documentElement;
  (Object.keys(set) as TokenName[]).forEach((key) => {
    root.style.setProperty(`--${key}`, set[key]);
  });
  validateContrast(set);
}
