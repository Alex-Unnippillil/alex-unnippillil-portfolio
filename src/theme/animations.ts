export const durations = {
  short: 100,
  medium: 200,
  long: 500,
};

export const easings = {
  standard: 'cubic-bezier(0.4, 0, 0.2, 1)',
  decelerate: 'cubic-bezier(0, 0, 0.2, 1)',
  accelerate: 'cubic-bezier(0.4, 0, 1, 1)',
  linear: 'linear',
};

export const applyAnimationVariables = (reduceMotion = false): void => {
  const root = document.documentElement;
  Object.entries(durations).forEach(([name, value]) => {
    const duration = reduceMotion ? 0 : value;
    root.style.setProperty(`--animation-duration-${name}`, `${duration}ms`);
  });
  Object.entries(easings).forEach(([name, value]) => {
    root.style.setProperty(`--animation-easing-${name}`, value);
  });
};

export default { durations, easings, applyAnimationVariables };
