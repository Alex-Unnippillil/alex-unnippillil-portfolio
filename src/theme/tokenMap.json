{
  "#232931": "var(--color-primary)",
  "#393e46": "var(--color-secondary)",
  "#4ecca3": "var(--color-active)",
  "#eee": "var(--color-background)",
  "#fff": "var(--color-white)",
  "#000": "var(--color-black)",
  "#b1b1b1": "var(--color-gray-400)",
  "#333": "var(--color-gray-800)",
  "#b2b2b2": "var(--color-gray-350)",
  "#e5e5e5": "var(--color-gray-200)",
  "#474a73": "var(--color-indigo-700)",
  "#3B5998": "var(--color-facebook)",
  "#181717": "var(--color-github)",
  "#E4405F": "var(--color-instagram)",
  "#0077B5": "var(--color-linkedin)",
  "#DA552F": "var(--color-producthunt)",
  "#FE7A16": "var(--color-stackoverflow)",
  "#1DA1F2": "var(--color-twitter)"
}
