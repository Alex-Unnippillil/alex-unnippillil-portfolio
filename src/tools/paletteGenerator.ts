export interface RGB {
  r: number;
  g: number;
  b: number;
}

export interface AccessiblePair {
  background: string;
  foreground: string;
}

function hexToRgb(hex: string): RGB {
  const sanitized = hex.replace('#', '');
  const bigint = parseInt(sanitized.length === 3 ? sanitized.split('').map((c) => c + c).join('') : sanitized, 16);
  return {
    r: (bigint >> 16) & 255,
    g: (bigint >> 8) & 255,
    b: bigint & 255,
  };
}

function luminance({ r, g, b }: RGB): number {
  const a = [r, g, b].map((v) => {
    const channel = v / 255;
    return channel <= 0.03928 ? channel / 12.92 : Math.pow((channel + 0.055) / 1.055, 2.4);
  });
  return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
}

export function contrast(color1: string, color2: string): number {
  const L1 = luminance(hexToRgb(color1));
  const L2 = luminance(hexToRgb(color2));
  const ratio = (Math.max(L1, L2) + 0.05) / (Math.min(L1, L2) + 0.05);
  return Math.round(ratio * 100) / 100;
}

export function generateAccessiblePair(background: string): AccessiblePair {
  const whiteContrast = contrast(background, '#ffffff');
  const blackContrast = contrast(background, '#000000');
  let foreground = whiteContrast >= blackContrast ? '#ffffff' : '#000000';

  if (contrast(background, foreground) < 4.5) {
    foreground = whiteContrast >= blackContrast ? '#ffffff' : '#000000';
  }

  return { background, foreground };
}

export default generateAccessiblePair;
