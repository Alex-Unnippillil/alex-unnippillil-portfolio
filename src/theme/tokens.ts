import { generateAccessiblePair, AccessiblePair } from '../tools/paletteGenerator';

export interface ThemeTokenMap {
  light: AccessiblePair;
  dark: AccessiblePair;
}

export const tokens: ThemeTokenMap = {
  light: generateAccessiblePair('#ffffff'),
  dark: generateAccessiblePair('#121212'),
};

export default tokens;
