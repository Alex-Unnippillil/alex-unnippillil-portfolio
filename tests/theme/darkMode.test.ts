import { tokens } from '../../src/theme/tokens';
import { contrast } from '../../src/tools/paletteGenerator';

describe('dark mode theme', () => {
  test('renders dark tokens', () => {
    expect(tokens.dark).toMatchSnapshot();
  });

  test('dark background and text have AA contrast', () => {
    const ratio = contrast(tokens.dark.background, tokens.dark.foreground);
    expect(ratio).toBeGreaterThanOrEqual(4.5);
  });
});
