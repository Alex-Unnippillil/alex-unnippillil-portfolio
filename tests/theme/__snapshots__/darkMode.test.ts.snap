// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark mode theme renders dark tokens 1`] = `
Object {
  "background": "#121212",
  "foreground": "#ffffff",
}
`;
